crates/typst/src/eval/version.rs

```diff
@@ -0,0 +1,223 @@
+        Self::COMPONENT_NAMES
+            .into_iter()
+            .enumerate()
+            .map(|(i, name)| (name, self.get(i).unwrap_or_default().get()))
+    }
+
+    /// Convert a version into an array
+    pub fn into_array(self) -> Array {
+        self.0.into_iter().map(|i| Value::Int(i.get())).collect()
+    }
+
+    pub fn push(&mut self, component: VersionComponent) {
+        self.0.push(component);
+    }
+}
+
+impl FromIterator<VersionComponent> for Version {
+    fn from_iter<T: IntoIterator<Item = VersionComponent>>(iter: T) -> Self {
+        Self(EcoVec::from_iter(iter))
+    }
+}
+
+impl Ord for Version {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let max_len = self.0.len().max(other.0.len());
+        let tail = repeat(&VersionComponent::ZERO);
+
+        let self_iter = self.0.iter().chain(tail.clone());
+        let other_iter = other.0.iter().chain(tail);
+
+        for (l, r) in self_iter.zip(other_iter).take(max_len) {
+            match l.cmp(r) {
+                Ordering::Equal => (),
+                ord => return ord,
+            }
+        }
+        Ordering::Equal
+    }
+}
+impl PartialOrd for Version {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
+}
+impl Eq for Version {}
+impl PartialEq for Version {
+    fn eq(&self, other: &Self) -> bool {
+        matches!(self.cmp(other), Ordering::Equal)
+    }
+}
+
+impl Hash for Version {
+    // Don't hash any explicitly suffixed zeros so that equal but non-identical versions (e.g. 1.2 and 1.2.0) hash as the same version
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        let len = self
+            .0
+            .iter()
+            .rposition(|x| x.get() != 0)
+            .unwrap_or_else(|| self.0.len());
+
+        self.0[..len].hash(state);
+    }
+}
+
+impl Display for Version {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        let mut first = true;
+        for &i in &self.0 {
+            if first {
+                first = false;
+            } else {
+                write!(f, ".")?;
+            }
+            write!(f, "{i}")?;
+        }
+        Ok(())
+    }
+}
+
+impl Debug for Version {
+    fn fmt(&self, f: &mut Formatter) -> std::fmt::Result {
+        write!(f, "version({})", self)
+    }
+}
+
+impl Serialize for Version {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        self.0.serialize(serializer)
+    }
+}
+
+#[derive(Default, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+#[repr(transparent)]
+pub struct VersionComponent(u64);
+
+#[allow(dead_code)] // reason: kept for completeness, might be used in the future
+impl VersionComponent {
+    pub const ZERO: Self = Self(0);
+
+    pub const MIN: Self = Self(0);
+    pub const MAX: Self = Self(i64::MAX as u64);
+
+    pub fn new(value: i64) -> Option<Self> {
```

Comment:
why is it exposed as a signed integer if it's really unsigned?

Reply:
Because typst uses signed integers for everything. I could've just as well made `new(value: u64)` and checked that the msb is zero, but that would make converting from typst int (i64) to VersionComponent one extra step, so I cut down on the redundancy.

Reply:
While Typst uses `i64` in the `Value` representation, the casting infrastructure allows you to use whatever type is most convenient and logical in any place. In this case, the best type would probably be `u32`. Perhaps, the whole `VersionComponent` type could be eliminated in favor of use `u32`.

Reply:
Hmmm, `u32` feels a bit limiting. It's probably not a real concern for version numbers, but I was trying to not let the max value be limited by virtue of being unsigned. Would you say u32 is "good enough" for versions?

Reply:
Hm, being able to deduplicate that would be really good imo. But let's discuss this in the main thread, not in the review comments.